File: sshx/src/providers.rs
use std::fmt;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Provider {
    GitHub,
    GitLab,
    Bitbucket,
    AzureDevOps,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum KeyType {
    Ed25519,
    Rsa4096,
}

impl Provider {
    pub const ALL: &[Provider] = &[
        Provider::GitHub,
        Provider::GitLab,
        Provider::Bitbucket,
        Provider::AzureDevOps,
    ];

    pub fn display_name(&self) -> &'static str {
        match self {
            Provider::GitHub => "GitHub",
            Provider::GitLab => "GitLab",
            Provider::Bitbucket => "Bitbucket",
            Provider::AzureDevOps => "Azure DevOps",
        }
    }

    pub fn default_hostname(&self) -> &'static str {
        match self {
            Provider::GitHub => "github.com",
            Provider::GitLab => "gitlab.com",
            Provider::Bitbucket => "bitbucket.org",
            Provider::AzureDevOps => "ssh.dev.azure.com",
        }
    }

    pub fn settings_url(&self) -> &'static str {
        match self {
            Provider::GitHub => "https://github.com/settings/ssh/new",
            Provider::GitLab => "https://gitlab.com/-/user_settings/ssh_keys",
            Provider::Bitbucket => "https://bitbucket.org/account/settings/ssh-keys/",
            Provider::AzureDevOps => "https://dev.azure.com/_usersSettings/keys",
        }
    }

    pub fn key_type(&self) -> KeyType {
        match self {
            Provider::AzureDevOps => KeyType::Rsa4096,
            _ => KeyType::Ed25519,
        }
    }

    #[allow(dead_code)]
    pub fn ssh_test_command(&self) -> &'static str {
        match self {
            Provider::GitHub => "git@github.com",
            Provider::GitLab => "git@gitlab.com",
            Provider::Bitbucket => "git@bitbucket.org",
            Provider::AzureDevOps => "git@ssh.dev.azure.com",
        }
    }

    pub fn from_str_loose(s: &str) -> Option<Provider> {
        match s.to_lowercase().as_str() {
            "github" | "gh" => Some(Provider::GitHub),
            "gitlab" | "gl" => Some(Provider::GitLab),
            "bitbucket" | "bb" => Some(Provider::Bitbucket),
            "azure" | "azuredevops" | "azure-devops" | "ado" => Some(Provider::AzureDevOps),
            _ => None,
        }
    }
}

impl fmt::Display for Provider {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.display_name())
    }
}

impl KeyType {
    pub fn algorithm(&self) -> &'static str {
        match self {
            KeyType::Ed25519 => "ed25519",
            KeyType::Rsa4096 => "rsa",
        }
    }

    pub fn bits(&self) -> Option<u32> {
        match self {
            KeyType::Ed25519 => None,
            KeyType::Rsa4096 => Some(4096),
        }
    }
}

File: src/providers.ts
export type ProviderKey = "bitbucket" | "github" | "gitlab" | "azure"

export type Provider = {
  label: string
  hostname: string
  settingsUrl: string
  keyType: "ed25519" | "rsa"
}

export const PROVIDERS: Record<ProviderKey, Provider> = {
  github: {
    label: "GitHub",
    hostname: "github.com",
    settingsUrl: "https://github.com/settings/keys",
    keyType: "ed25519",
  },
  gitlab: {
    label: "GitLab",
    hostname: "gitlab.com",
    settingsUrl: "https://gitlab.com/-/profile/keys",
    keyType: "ed25519",
  },
  bitbucket: {
    label: "Bitbucket",
    hostname: "bitbucket.org",
    settingsUrl: "https://bitbucket.org/account/settings/ssh-keys/",
    keyType: "ed25519",
  },
  azure: {
    label: "Azure DevOps",
    hostname: "ssh.dev.azure.com",
    settingsUrl: "https://dev.azure.com/_usersSettings/keys",
    keyType: "rsa",
  },
}

export const PROVIDER_KEYS = Object.keys(PROVIDERS) as ProviderKey[]

export function detectProviderFromHostname(hostname: string): ProviderKey | undefined {
  for (const [key, provider] of Object.entries(PROVIDERS) as [ProviderKey, Provider][]) {
    if (hostname.includes(provider.hostname)) return key
  }
  return undefined
}

export function detectProviderFromRemoteUrl(url: string): ProviderKey | undefined {
  if (url.includes("github.com")) return "github"
  if (url.includes("gitlab.com")) return "gitlab"
  if (url.includes("bitbucket.org")) return "bitbucket"
  if (url.includes("dev.azure.com") || url.includes("vs-ssh.visualstudio.com")) return "azure"
  return undefined
}

export function normalizeProviderKey(value: string): ProviderKey | undefined {
  const v = value.trim().toLowerCase()
  if (v === "github" || v === "gh") return "github"
  if (v === "gitlab" || v === "gl") return "gitlab"
  if (v === "bitbucket" || v === "bb") return "bitbucket"
  if (v === "azure" || v === "azure-devops" || v === "azuredevops") return "azure"
  return undefined
}
